### ISSUE TITLE
Specify machine to build

### ISSUE THREAD

**Issue by larsbrinkhoff** [PDP-10/its#1323]
Add a way to specify which machine to build.  E.g.

    make all MCHN=LC

This will affect the build script for:

- ITS
- SALV
- DUMP
- MAGDMP
- MAGFRM

**Comment by larsbrinkhoff** [comment]
The classic four would have differents sets of software installed:

- AI: Type 340, PDP-6, TV, XGP, Logo, GT40, SUDS, Maclisp, MacHack VI, Lispm
- DM: Muddle, demons, Imlac, Maze
- MC: Maclisp, Macsyma
- ML: ?

This division already maps somewhat to our build scripts: lisp.tcl, macsyma.tcl, muddle.tcl.  Some more could be split off from misc.tcl suggested by the above.

**Comment by larsbrinkhoff** [comment]
Build scripts to split off:

- 340.tcl
- pdp6.tcl
- tv.tcl
- gt40.tcl
- imlac.tcl
- xgp.tcl
- logo.tcl

**Comment by larsbrinkhoff** [comment]
Or maybe just ai.tcl, dm.tcl, mc.tcl?

**Comment by eswenson1** [comment]
Not sure I like this suggestion — at least not if the build is exclusively one (machine) or another. For example, those of us intersected in iTS generally, and not recreating some old historical environment, such as AI KA, want to be able to have as many ITS programs as possible — that run at all on the generic combination environment. I wouldn’t want to have to build an MC to get Macsyma, and in doing so, not get LLOGO. 

I think adding options to the build to get specific features is fine, but having a single “machine” designator — like “ai” is wrong.

There is even value, for example, in an environment without an 11 or Knight TV, to have programs that use these interfaces — if for no other reason for exploration/code perusal purposes. 

Again, I like having build flags that allow including everything (buildable for a given emulator) in a build. Those of us who want everything can get it. Those who don’t can get selective pieces, as well.

**Comment by larsbrinkhoff** [comment]
The default should remain the inclusive build as now.

There are two ideas behind this:
- I'd like the ITS name to be more flexible, i.e. to be able to build NO ITS.
- In a few cases different programs had the same names.  For example WHOM means something else on DM.

**Comment by eswenson1** [comment]
I certainly agree it should be possible to configure the host name and network info.  And it should be possible to build specific configurations (where parameters can pick and choose or define sets of components based on parameters).  I'm not sure whether the WHOM case is a very good example of the need to do things differently for various machines. Why not pick one of them (the one for AI/MC/ML) and make that the default and rename the other one WHOMDM or something like that?

I like what we did with macsyma -- where you can conditionally include it in a the build.  I agree we should follow a similar model, with build flags selecting major pieces of functionality, as long as the default build gets everything possible (as long as what is included is reasonable for the underlying emulator).

**Comment by larsbrinkhoff** [comment]
Yes, I think I named it WHOMUD.  But I don't think this is entirely fair to DM ITS.  If there were a way to target DM specifically, I'd like to have it named WHOM there.  But it would still be named WHOMUD in the default build.

DM ITS has many other differences that I think may not be appropriate to fold into the default ITS build.  Special ITS system calls, and a different set of demons.

**Comment by eswenson1** [comment]
Yes, historically DM was a very separate beast.  We regularly updated ITS and system programs (using INSTALL) on MC/AI/ML, but never touched DM.

**Comment by larsbrinkhoff** [comment]
Thank you.  That confirms the impression I had.

I have a fairly good idea what kind of work was done on AI and MC.  There's less information about DM due to lack of archives files; still, we know about Muddle, Imlacs, games, etc.  I actually have no idea about ML.

**Comment by larsbrinkhoff** [comment]
NETRFC works differently for DM and other ITS machines.  #1475

**Comment by larsbrinkhoff** [comment]
The DIS device can be either the 340 or the LDS-1.  But not both; they are mutually exclusive.  For one thing, they both use the device name "DIS", and another, device code 130.

So if or when there is a simulation of the LDS-1, there must be a way for the user to select between a 340 or LDS-1 display.

**Comment by larsbrinkhoff** [comment]
@eswenson1 and I discussed a more flexible build system.  The current system can only build a fixed configuration (except some minor parameters in "conf") for each particular PDP-10 emulator.

A more flexible system should parametrize these:
- Emulator (as before).
- Processor: KA10, KL10, or KS10.  This depends on the emulator used.
- Disk controller: DC10, RP10, RH10, or RH11.
- Disk packs, type and how many: Memorex, Calcomp, RP03, RP04, RP06, ...
- Tape controller: TM10A, TM10B, or TM03.
- System name: AI, DM, ML, MC, etc.

**Comment by larsbrinkhoff** [comment]
My preference for this would be to make small incremental steps rather than a huge rewrite from scratch.

Possibly, but maybe not as the first step, the user would just specify emulator and system.  The hardware parameters could be parsed out from the SYSTEM; CONFIG > file.

**Comment by larsbrinkhoff** [comment]
The most pressing issue is to build with another disk controller.

I imagine these items would be split off from the scripts to separate files:
- Calling MARK from SALV.
- SALV assembly-time switches.
- DSKDMP assembly-time switches.
- Calling UCOP from SALV.

And something to update CONFIG and emulator scripts to match.

**Comment by larsbrinkhoff** [comment]
I suggest AI, DM, ML, and MC should be reserved for hosts on a simulated ARPANET.

**Comment by larsbrinkhoff** [comment]
A small step has been implemented.  The make command line will accept MCHN= to set the ITS name, see #1898.

**Comment by larsbrinkhoff** [comment]
This is now more urgent due to the possibility of reenacting ARPANET hosts.
